### ISSUE TITLE
c-unwind fail on nightly

### ISSUE THREAD

**Issue by thewh1teagle** [rust-lang/rust#127651]
Hello!
I'm trying to use `c-unwind` feature so I can catch `abort()` call from `c` / `c++`
However, seems like it doesn't catch it. I tried on `macOS` / `Linux` and `Windows`

I tried this code:

```rust
extern "C-unwind" {
    fn printHelloAndAbort();
}

fn main() {    
    let result = unsafe {
        std::panic::catch_unwind(|| {
            printHelloAndAbort();
        })
    };

    if let Err(_) = result {
        println!("Caught a panic from C");
    }
}
```

I expected to see this happen: *explanation*
I expected that it should print `Catch a panic from C`

Instead, this happened: *explanation*
Ubuntu:
```console
/home/runner/work/_temp/6763a687-7c6c-4e26-b151-8c6aba9fc683.sh: line 1:  1919 Aborted                 (core dumped) cargo run
```
Windows:
```console
error: process didn't exit successfully: `target\debug\c-unwind.exe` (exit code: 0xc0000409, STATUS_STACK_BUFFER_OVERRUN)
```

macOS:
```console
Hello from C
/Users/runner/work/_temp/f3bdf460-5318-4be2-98[7](https://github.com/thewh1teagle/c-unwind/actions/runs/9906658173/job/27368669558#step:5:8)b-3f6ddba5bcfb.sh: line 1:  2339 Abort trap: 6           cargo run
```

### Meta

`rustc --version --verbose`:
```
rustc 1.81.0-nightly (524d806c6 2024-07-05)
binary: rustc
commit-hash: 524d806c62a82ecc0cf8634b94997ae506f4d6f9
commit-date: 2024-07-05
host: x86_64-pc-windows-msvc
release: 1.81.0-nightly
LLVM version: 18.1.7
```

<details><summary>Backtrace</summary>
<p>

```

```

</p>
</details>

Repro source code:
https://github.com/thewh1teagle/c-unwind

Github actions logs:
https://github.com/thewh1teagle/c-unwind/actions/runs/9906658173

If I'm wrong and that's not what `catch_unwind` should do, is there another way I can catch abort() in c bindings?
Thanks

**Comment by eggyal** [comment]
1. `abort()` doesn't unwind: it just immediately aborts.
2. `catch_unwind` can only be used to catch Rust panics, not foreign ones.

**Comment by thewh1teagle** [comment]
> 1. `abort()` doesn't unwind: it just immediately aborts.  Unlike C++ (that has exceptions), C doesn't have any language support for stack unwinding (though one can of course implement it or use a library that does so).

Thanks for explaining. I understand that we can't unwind from c abort since it doesn't unwind. But regarding c++, I didn't fully understand. I understand that c++ 's abort does support unwinding, so can we catch it with catch_unwind on c++ function? Because I tried the same code on a c++ function, and catch_unwind didn't work there either.

**Comment by eggyal** [comment]
C++'s `abort()` is the same as C's: it does not unwind.  `throw`ing an exception in C++ causes the stack to unwind, though as mentioned in my previous comment `catch_unwind` can only be used to catch Rust panics, not foreign exceptions.

**Comment by eggyal** [comment]
FWIW, [URLO](https://users.rust-lang.org) may be a better forum for this discussion.
